src/python/pants/conftest.py

```diff
@@ -42,3 +44,35 @@ def pytest_sessionfinish(session) -> None:
     # Technically unecessary, but nice if people are running tests directly from repo
     # (not using pants).
     namespace_init_path.unlink()
+
+
+@pytest.fixture(autouse=True, scope="session")
+def dedicated_target_fields():
+    """Ensures we follow our convention of dedicated source and dependencies field per-target.
+
+    This help ensure that plugin authors can do dependency inference on _specific_ field types, and
+    not have to filter targets using generic field types.
+
+    Note that this can't help us if a target type should use an _even more specialized_ dependencies
+    field type (E.g. 100 different target subclasses could use 1 custom dependencies field type,
+    when in reality there should be many more). However, this is a good sanity check.
+    """
+    from pants.engine.target import Dependencies, SourcesField, Target
+
+    for cls in Target.__subclasses__():
+        if hasattr(cls, "core_fields"):
+            for field_cls in cls.core_fields:
+                if (
+                    issubclass(field_cls, SourcesField)
+                    and field_cls.__module__ is SourcesField.__module__
+                ):
```

Comment:
Why not simply?
```python
if field_cls is SourcesField:
    raise ...
```

Reply:
We're not just checking for `SourcesField`, there's `SingleSourceField`, `MultipleSourcesField`, `OptionalSingleSourceField` too

Reply:
FWIW I tried adding ` conftest_test.py` testing this behavior, but since this is done at session scope, it is not so easy to test :disappointed:

Reply:
Ooh... ok, cool. Kind of a gotcha that it's sneaky. Would consider `if field_cls in (SourcesField, SingleSourcesField, Multi.....)` but then that's list that needs maintaining, so not sure it improves matters then.. perhaps a comment to highlight this nuance is better?